PS-4289: PXC 5.7.43 refresh - Q3 2023 (merge part 2)

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-4289

Testing Done
---
Jenkins: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/394/

Commit:
Bug#34711762 Upgrade zlib to 1.2.13 [add tarball]

Unpack source tarball, git add everything.

Change-Id: Ib6eb64f8e132ca59539208f7bf69245268804ee5

Commit:
Bug#34711762 Upgrade zlib to 1.2.13 [remove unneeded]

Remove things we do not need/want.

git rm -rf amiga/ contrib/ doc/ examples/ nintendods/ Makefile zconf.h

Change-Id: Ibd76884411c6596f2fcfcb6c3fe2f1f4aabadb73

Commit:
Bug#34711762 Upgrade zlib to 1.2.13 [cmake changes]

Bump MIN_ZLIB_VERSION_REQUIRED to "1.2.13"
and adjust paths to bundled zlib sources.

In extra/zlib/zlib-1.2.13/CMakeLists.txt:
 - apply cumulative patches from previous zlib upgrade

Change-Id: I1a0aeff115a96a0993f2f396c643eda1c1b4900b

Commit:
Bug#34711762 Upgrade zlib to 1.2.13 [remove old]

Remove all old source files.

Change-Id: I456635823feb21faa42b683f0bfb62d353cb80d4

bkandasa committed:
Raise version number after cloning 5.6.59

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.41

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

TharunSai committed:
Bug #34857411 : regression - slow connections/telnet block many other statements
                for connect_timeout seconds, causing pileups

Description:
------------
This is a regression caused due to the fix made for the Bug 34094706. When a
connection somehow stalls/blocks during the authentication phase, where a mutex
is held, the other connections that are executing queries on I_S and P_S are
blocked until the first connection release the mutex.

Fix:
----
Instead of using the mutex and checking the thd->active_vio, we now check the
value of net.vio type in the is_secure_transport() check.

Change-Id: Ic301a3a69b2bb828f2323441bdd0232e5a0ecef4

shussinh committed:
BUG# 34574013 - backport 33723597 to 5.7

Backporting the bug 33723597 to 5.7

A connect packet can read in chunks. But the connect_timeout applies
to reading the individual chunks. And not to the full packet being read.
This can lead to very long reads in the client is taking its time to
write the full packet.
Fixed by also enforcing connect_timeout on the full packet read and not
only to the individual chunks.

Additional Changes:
In previous code base, it was only the socket connection types which
had server extension (ext) along with them, other connection types
such as named pipe, shared memory etc. didn't have them.

In order to fetch the timeout_on_full_packet from the NET_SERVER
for named pipe, and shared memory connections, we need to have this
extension available to all connection types, not just to socket
connections.

Moved the "init_net_server_extension" from socket_connection.cc to a
common header and source file. Added the function calls to
named_pipe_connection.cc as well as shared_memory_connection.cc, and
modified socket_connection.cc accordingly.

Change-Id: Ibbaff8e26445b04d04f4e812be561eefe150157f

inikep committed:
PS-8820 [5.7]: Add gcc-13 and clang-16 to Azure Pipelines and fix gcc-13 compilation issues

1. Add gcc-13 and clang-16 to Azure Pipelines
2. Update RocksDB submodule pointer

-------------------------------------------------------------------------------

[cmake] Bug #109325: Ubuntu 22.04 forces LTO compilation with gcc in RelWithDebInfo

Remove LTO flags from `dpkg-buildflags`.

Ubuntu 22.04 Jammy returns:
```
$ dpkg-buildflags --get CXXFLAGS
-g -O2 -ffile-prefix-map=/data/mysql-server=. -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security
```

Ubuntu 20.04 returns:
```
$ dpkg-buildflags --get CXXFLAGS
-g -O2 -fdebug-prefix-map=/data/docker=. -fstack-protector-strong -Wformat -Werror=format-security
```

After this patch the result for Ubuntu 22.04 matches Ubuntu 20.04.

Commit:
Bug#34617506: Corrupted replication event may lead mysql coredump

Problem:
Server crashes when applying a corrupted Rows log event(Update rows log event).

Analysis:
If an error happens(due to corrupted event) while reading the BI(before image)
when unpack_current_row was called in the method do_index_scan_and_update,
then it sets m_curr_row_end to NULL.
Later in the same method we assign this value NULL to m_curr_row and call
the unpack_current_row(to read the After Image), that casues the server to crash.

Fix:
Don't try to read the After Image through unpack_current_row method in case
we encounter an error while reading Before Image and m_curr_row_end is NULL.

Change-Id: I489dd20ae37d37bcb3bdf199d01050503aa6d3b6

Commit:
Bug#34750489: applying a updated online log may access out-of-bounds memory

Some online DDL operations scenes, which need rebuilding the table, may
access out-of-bounds memory, when the table has virtual columns.

row_log_table_apply_op() needs to check buffer length before calls
trx_undo_read_v_cols().

In addition, 8.0 code around this fix seems to have problem when mrec_end==next_mrec case.
(might miss to call trx_undo_read_v_cols() wrongly and cause mismatch log reading.)
The 8.0 code should be also same with this 5.7 fixed code.

Change-Id: I016a581b0a6fd122921c9d4c04e655ceec679ed1

Commit:
Bug#34741689: mysqld crash at row_upd_sec_index_entry_low

row_upd_store_v_row() might call handler::my_eval_gcolumn_expr().
But it might overwrite new_row buffer from mysqld
and v_col's node->update->fields[*].new_val.data contents might be changed.

So, before calls handler::my_eval_gcolumn_expr(),
v_col's node->update->fields[*].new_val should be duplicated.

In addition, 8.0 code needs to treat also multi value fields.
Treating with the duplicate code as same as just before this fix.

Change-Id: I95d5c18727d5ae6ff8a9bf501164339f1f03f18d

Commit:
BUG#34930969 Contribution by Tencent: Previous_gtids miss gtid when binlog_order_commits off

Problem
-------

GTID may be missed in next binlog file's Previous_gtids event if binlog_order_commits off.

Analysis / Root-cause analysis
------------------------------

When binlog_order_committs is turned off, binlog rotate will wait for 'm_prep_xids' decrease
to zero. However, 'm_prep_xids' is decreased before the update of thd's GTID to 'gtid_executed'
(see function (see function MYSQL_BIN_LOG::finish_commit).

Solution
--------

Call dec_prep_xids after gtid_state->update_on_commit in
MYSQL_BIN_LOG::finish_commit

Change-Id: Ic0855d8b6f962b43aeb06180e6f60f28a5780d5a

gipulla committed:
Merge branch 'mysql-5.7.41-release' into mysql-5.7

Commit:
BUG#34750847: UPDATE COPYRIGHT HEADERS TO 2023 [5.7 patch]

Update all copyright headers in 5.7 to 2023 and to the current
copyright format.  This patch was generated by a script.

Approved-by: terje.rosten@oracle.com
Change-Id: Ia9481f760ca25b591893d2442dca07106ac9c3cb

Commit:
Post-push fix: BUG#34750847: UPDATE COPYRIGHT HEADERS TO 2023

Change-Id: Ia501181cf77b018e3cc661c6503543f533482ac5

shussinh committed:
BUG# 35020512 - Provide an option in mysqldump to disable
FLUSH TABLES WITH READ LOCK so that RELOAD privilege is
not required [mysql-5.7]

Problem:
-------
Starting mysqldump with --single-transaction is throwing the below
error for users without the RELOAD privilege:

mysqldump: Couldn't execute 'FLUSH TABLES': Access denied; you need
(at least one of) the RELOAD privilege(s) for this operation (1227)

Analysis:
---------
The changes made in the bug fix of 33630199 caused a regression
in which whenever user passes --single-transaction, mysqldump
tries to perform FLUSH TABLES WITH READ LOCK.

Now "FLUSH TABLES" operation requires the users to have the RELOAD
privilege, and many dump users do not have this privilege.

This FLUSH TABLES was required in the case of bug 33630199 where
it had issues with GTID consistency, and FLUSH TABLES was solving
this issue, but the FLUSH TABLES is not required for every dump
started with --single-transaction. It is needed only when the
server has GTIDs enabled and --set-gtid-purged is set to ON or AUTO.

This is how it was checking if FLUSH TABLES is needed, previously:

  if ((opt_lock_all_tables || opt_master_data ||
      (opt_single_transaction && flush_logs)) &&
      do_flush_tables_read_lock(mysql))
    goto err;

after the bug fix of 33630199, it was changed to this

  if ((opt_lock_all_tables || opt_master_data ||
       opt_single_transaction) &&
       do_flush_tables_read_lock(mysql))
    goto err;

and this, performs FLUSH TABLES every time the user
passes --single-transaction.

Fix:
----
The fix is to perform FLUSH TABLES only when GTIDs are enabled
on the server, AND --set-gtid-purged is set to ON /or/ AUTO

No need to perform FLUSH TABLES for every dump started
with --single-transaction.

Restored the previous if block, with one additional check
for condition mentioned above.

Change-Id: I08bcf3813b6a9d405863c1e9f667c85b3e059288

Commit:
Bug#35095010 CURL UPDATE TO AT LEAST 7.87.0 / 7.88.0 [noclose]

Patch for 5.7

There is only one CURL client in 5.7, and it is not affected by the
upgrade.

Verified by backporting CURL cmake code from trunk,
and setting proper include paths for CURL headers.

Change-Id: I888507e72c297c63f77ce9d90df79b08b6d38d59

bkandasa committed:
Update License Book

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.42

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

Commit:
Bug#35158675 Compile MySQL 5.7 with gcc 12

Compiling MySQL 5.7 with gcc 12, we need to backport some cmake code:
Disable -Wunused-label when compiling the generated BISON parsers.

Note that 5.7 is not officially supported for gcc 12,
this is just for developer convenience.

Change-Id: I8c91fb359703cbc667eb641aa0aa725743d7b738

TharunSai committed:
Bug #34857411 : regression - slow connections/telnet block many other statements
for connect_timeout seconds, causing pileups

[post-push-fix] Fixing the sporadically failing testcase on PB2

Approved by Harin Vadodaria <harin.vadodaria@oracle.com> over Slack.

TharunSai committed:
Bug #34857411 : regression - slow connections/telnet block many other statements
for connect_timeout seconds, causing pileups

[post-push-fix] Fixing the sporadically failing testcase on PB2

Approved by Harin Vadodaria <harin.vadodaria@oracle.com> over Slack.

(cherry picked from commit 5bbbeeb95a0f17065a69a02984b1fc8c5d54f693)

vsmurthys committed:
Bug#35054579 Issue in Oracle MySQL Client using utf16 charset

Back porting the bug to 5.7 branch.

Change-Id: I06810ab5d85ac6f65b94ff1d12bc116cf40d7fa1
(cherry picked from commit 4b39338dfc1fba298e80812b5e009906040beea5)

vsmurthys committed:
Bug#35054579 Issue in Oracle MySQL Client using utf16 charset

Back porting the bug to 5.7 branch.

Change-Id: I06810ab5d85ac6f65b94ff1d12bc116cf40d7fa1

mbremyk committed:
Bug#32824429 - SIGNAL 11 / VIRTUAL LONGLONG ITEM_FUNC_IN::VAL_INT():
		ASSERTION `IN_ITEM'

Problem:
	Floating point literal in query was identified as real number,
	then converted to zerofilled string to compare with
	zerofilled float column.
	Due to being identified as a real number the comparison was
	attempted in a real context, even after the value was converted
	to string.
Solution:
	Convert zerofill numbers to zerofilled Item_string only when
	used in a string context, eg. concat or comparing with a string
	constant.

(cherry picked from commit ce4d1d1b605535db3f00bedc810bce93e7b7ccc0)
Change-Id: Id1e3ddcb917471258ca5e003e075fcb0b4a09b21

tiagoportelajorge committed:
Bug#34976442 mysqld crashed

Problem
====================

On an MySQL Innodb cluster deployment which has:
- A virtualization platform of Docker/Kuberenetes;
- 3 nodes working and running

During regular operation, one of the nodes crashes and it is was unable
to start and join the group again. In order to restore the system,
the whole group was restarted and then all members crashed in
get_ipv4_local_addresses within GCS/XCom.

Analysis and Fix
====================

Docker/Kuberenetes create a lot of network interfaces within a pod,
for a virtual network. In this specific case, more than 64 interfaces.

What happened was that XCom allocates an array of 64 interfaces which
is reallocated each time we overflow it. Specifically,
`ifrp` is overflowed when network interface is more than
IFRP_INIT_ARR_SIZE(64). But it is not reallocated because wrong
condition `i==ifrpsize`.

The solution is to fix the wrong reallocate condition.

Change-Id: Ibbf150e4f6a7fc84db8d44193c3f729aba7d7089

msprajap committed:
Merge branch 'mysql-5.7.42-release' into mysql-5.7

zmur committed:
WL#15614 OpenSSL 3.0 support in Cluster v7.5, 7.6/MySQL 5.7

- adapt finding OpenSSL version from include file for OpenSSL 3.0
- allow use of OpenSSL version 3
- suppress deprecation warnings for using deprecated OpenSSL API
  - in a different way for rapid/unittest due to build problems with
    GNU 4.4.7 compiler.
- allow multiple definitions for symbols both in libssl and libcrypto
  when building against OpenSSL 3.0.8, see
  https://github.com/openssl/openssl/issues/20238.
- fix failing check in debug in vio (bug#33967647)
- adapt test x.connection_tls_version to slight change in errors
  reported from OpenSSL 3.0
- provide openssl.cnf for MTR tests that explicitly tests TLSv1.0 and
  TLSv1.1 support.

Change-Id: Ic82918b61b2bfa6e6188bad3c716950bd909c503

gkodinov committed:
Bug#35374491: Mysql read_ok_ex missing length check

Added a validity check on the length received from the network
for unknown trackers.

Change-Id: Idc7d229fc5905d919bd8c59907be1a5439a08efe

Commit:
Bug#34758905 Range analysis for IN list has O(n^2) runtime complexity

During optimization, range SEL_TREE creation uses different logic
based on the LHS of the IN predicate. For a field item, each RHS
value is added to an OR tree to create the necessary expression. In
the case of a row item comparison, a DNF expression is needed. A DNF
expression is created by adding an AND tree with column values to an
OR  tree for each set of RHS values. Instead of adding each new AND
tree to the OR tree, the OR tree is added to the AND tree causing
exponential time during the tree merge.

Solution: Change the tree joining the order. Instead of adding OR
tree to the AND tree, add AND tree to the OR tree.

Change-Id: I93503f3a8e3d6fd181aa40364d24f2868f294168

bjornmu committed:
Update License Book

Approved-by: Erlend Dahl <erlend.dahl@oracle.com>

Commit:
PS-8785: Increment metrics for the first iteration of buf_LRU_free_from_common_LRU_list()

https://jira.percona.com/browse/PS-8785

Currently buffer_LRU_search_scanned and buffer_LRU_unzip_search_scanned
metrics are not incremented by buf_LRU_free_from_common_LRU_list() and
buf_LRU_free_from_unzip_LRU_list() methods in case the first item from
the list gets evicted.

- Updated logic in buf_LRU_free_from_common_LRU_list() and
  buf_LRU_free_from_unzip_LRU_list() to take into account first item in
  unzip_LRU and LRU lists.
- Removed now redundant "!freed" condition in 'for' loop condition.

adivinho committed:
Merge pull request #5061 from percona/release-5.7.42-46

PS-8779 Release tasks ticket for PS 5.7.42-46

bjornmu committed:
Update License Book

Approved-by: Balasubramanian Kandasamy <balasubramanian.kandasamy@oracle.com>
(cherry picked from commit eee3f0851c4942a11b2e55818d51484de016fcd4)

oleksandr-kachan committed:
Merge pull request #5067 from oleksandr-kachan/PS-8785-5.7

PS-8785: Increment metrics for the first iteration of buf_LRU_free_from_common_LRU_list() (5.7)

inikep committed:
Merge pull request #5076 from inikep/PS-8820-5.7

PS-8820 [5.7]: Add gcc-13 and clang-16 to Azure Pipelines and fix gcc-13 compilation issues

Commit:
Implemented PS-8843 (Merge MySQL 5.7.43) (docs)

https://jira.percona.com/browse/PS-8843
***
Updated man pages from MySQL Server 5.7.43 source tarball.
***
Updated 'scripts/fill_help_tables.sql' from MySQL Server 5.7.43 source tarball.

percona-ysorokin committed:
This is a "fixing" mysql-5.7.42 -> percona-server 5.7 null merge

In Q3 2023 Oracle released MySQL Server 5.7.43 and 8.0.34 and partially
re-wrote git commit history in their 5.7 git branch (about 30
'--first-parent' commits).
This "fixing" null merge is supposed to bring new Oracle mysql-server commit
hashes (up to mysql-server-5.7.42) into existing percona-server 5.7 branch.

This merge was done via
git merge -s ours mysql-5.7.42

Commit:
Implemented PS-8843 (Merge MySQL 5.7.43 up to 00173cb60cd) (merge with conflicts)

https://jira.percona.com/browse/PS-8843

Commit:
Implemented PS-8843 (Merge MySQL 5.7.43 up to 00173cb60cd) (conflicts resolved)

https://jira.percona.com/browse/PS-8843
***
VERSION raised to "5.7.43-47.0".
PERCONA_INNODB_VERSION in univ.i raised to "47.0".

Commit:
Implemented PS-8843 (Merge MySQL 5.7.43) (no conflicts)

https://jira.percona.com/browse/PS-8843

oleksandr-kachan committed:
Merge pull request #5084 from oleksandr-kachan/merge-mysql-5.7.43-part1

PS-8843: Merge MySQL 5.7.43 (part 1)

oleksandr-kachan committed:
Merge pull request #5090 from oleksandr-kachan/merge-mysql-5.7.43-part2

PS-8843: Merge MySQL 5.7.43 (part 2)

adivinho committed:
PS-8875 Release tasks ticket for PS-5.7.43-47

adivinho committed:
Merge pull request #5097 from adivinho/release-5.7.43-47

PS-8875 Release tasks ticket for PS-5.7.43-47

Commit:
PS-4289: PXC 5.7.43 refresh - Q3 2023 (merge part 1 with conflicts)

https://jira.percona.com/browse/PXC-4289

Merge commit '869fb3c' into merge-5.7.43-part1

***
Resolved .gitmodules conflict to be able to commit.

Conflicts:
        .gitmodules
        MYSQL_VERSION
        man/ndb_config.1
        storage/rocksdb/rocksdb

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1834) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1834) before we can accept your contribution.<br/>**3** out of **15** committers have signed the CLA.<br/><br/>:white_check_mark: adivinho<br/>:white_check_mark: inikep<br/>:white_check_mark: venkatesh-prasad-v<br/>:x: bkandasa<br/>:x: gipulla<br/>:x: shussinh<br/>:x: msprajap<br/>:x: Sai Tharun Ambati<br/>:x: gkodinov<br/>:x: bjornmu<br/>:x: percona-ysorokin<br/>:x: zmur<br/>:x: tiagoportelajorge<br/>:x: mbremyk<br/>:x: oleksandr-kachan<br/><hr/>**Sai Tharun Ambati** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtradb-cluster?pullRequest=1834) it.</sub>

Commit:
PS-4289: PXC 5.7.43 refresh - Q3 2023 (merge part 1 conflicts resolved)

https://jira.percona.com/browse/PXC-4289

Commit:
PS-4289: PXC 5.7.43 refresh - Q3 2023 (merge part 2 no conflicts)

https://jira.percona.com/browse/PXC-4289

Merge tag 'Percona-Server-5.7.43-47' into merge-5.7.43-part2

Percona Server release 5.7.43-47

Commit:
PS-4289: PXC 5.7.43 refresh - Q3 2023 (Fix MTR tests)

As a follow-up with the bugfix
(PXC-4218: Truncated output of SHOW VARIABLES LIKE 'wsrep_provider_options')
(commit ac2a83e46)
that increased the buffer size for server variables to 4096 chars, the
following tests were re-recorded:

1. perfschema.table_schema
2. x.notices_enable

kamil-holubicki commented:
As we discussed on Slack, let's make it compilable with gcc-13 as well

Commit:
PS-4289: PXC 5.7.43 refresh - Q3 2023 (Fix gcc-13 errors)

https://jira.percona.com/browse/PXC-4289

Description
-----------
Server compilation fails with gcc-13 due to nonnull-compare of `this`
pointer in the WSREP(THD) macro.

  sql/sql_class.h: In member function ‘void THD::set_query_id(query_id_t, bool)’:
  sql/sql_class.h:5196:9: error: ‘nonnull’ argument ‘this’ compared to NULL [-Werror=nonnull-compare]

We get this error when `this` pointer is compared with NULL when
Werror=nonnull-compare is specified.

Solution
--------
Occurences of `WSREP(this)` have been replaced with `WSREP_NNULL(this)`.

kamil-holubicki reviewed (approved):
LGTM. Please remember to merge Part#1 first :)
